# **Kaggle – DataTops®**
Tu TA ha decidido cambiar de aires y, por eso, ha comprado una tienda de portátiles. Sin embargo, su única especialidad es Data Science, por lo que ha decidido crear un modelo de ML para establecer los mejores precios.

¿Podrías ayudar a tu profe a mejorar ese modelo?

## Aspectos importantes
- Última submission:
    - Mañana: 17 de febrero a las 5pm
    - Tarde: 19 de febrero a las 5pm
- **Enlace de la competición**: https://www.kaggle.com/t/c5cc87b50c4b4770bdc8f5acbe15577d
- **Requisito**: Estar registrado en [Kaggle](https://www.kaggle.com/)

## Métrica:
El error cuadrático medio (RMSE, por sus siglas en inglés) es una medida de la desviación estándar de los residuos (errores de predicción). Los residuos representan la diferencia entre los valores observados y los valores predichos por el modelo. El RMSE indica qué tan dispersos están estos errores: cuanto menor es el RMSE, más cercanas están las predicciones a los valores reales. En otras palabras, el RMSE mide qué tan bien se ajusta la línea de regresión a los datos.


$$ RMSE = \sqrt{\frac{1}{n}\Sigma_{i=1}^{n}{\Big(\frac{d_i -f_i}{\sigma_i}\Big)^2}}$$


## 1. Librerías

In [17]:
import numpy as np
import pandas as pd
from PIL import Image
from sklearn.model_selection import train_test_split
from sklearn.metrics import root_mean_squared_error
from sklearn.ensemble import RandomForestRegressor
from sklearn.svm import SVR
import urllib.request

## 2. Datos

In [18]:
# Para que funcione necesitas bajarte los archivos de datos de Kaggle
df = pd.read_csv("./data/train.csv")

### 2.1 Exploración de los datos

In [19]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 912 entries, 0 to 911
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   laptop_ID         912 non-null    int64  
 1   Company           912 non-null    object 
 2   Product           912 non-null    object 
 3   TypeName          912 non-null    object 
 4   Inches            912 non-null    float64
 5   ScreenResolution  912 non-null    object 
 6   Cpu               912 non-null    object 
 7   Ram               912 non-null    object 
 8   Memory            912 non-null    object 
 9   Gpu               912 non-null    object 
 10  OpSys             912 non-null    object 
 11  Weight            912 non-null    object 
 12  Price_in_euros    912 non-null    float64
dtypes: float64(2), int64(1), object(10)
memory usage: 92.8+ KB


In [20]:
df.head()

laptop_ID Company        Product   TypeName  Inches   ScreenResolution  \
0        755      HP         250 G6   Notebook    15.6  Full HD 1920x1080   
1        618    Dell  Inspiron 7559     Gaming    15.6  Full HD 1920x1080   
2        909      HP    ProBook 450   Notebook    15.6  Full HD 1920x1080   
3          2   Apple    Macbook Air  Ultrabook    13.3           1440x900   
4        286    Dell  Inspiron 3567   Notebook    15.6  Full HD 1920x1080   

                           Cpu   Ram               Memory  \
0     Intel Core i3 6006U 2GHz   8GB            256GB SSD   
1  Intel Core i7 6700HQ 2.6GHz  16GB              1TB HDD   
2   Intel Core i7 7500U 2.7GHz   8GB              1TB HDD   
3         Intel Core i5 1.8GHz   8GB  128GB Flash Storage   
4   Intel Core i3 6006U 2.0GHz   4GB              1TB HDD   

                              Gpu       OpSys  Weight  Price_in_euros  
0           Intel HD Graphics 520  Windows 10  1.86kg          539.00  
1  Nvidia GeForce GTX 960<U+039C>  Windows 10  2.59kg          879.01  
2            Nvidia GeForce 930MX  Windows 10  2.04kg          900.00  
3          Intel HD Graphics 6000       macOS  1.34kg          898.94  
4              AMD Radeon R5 M430       Linux  2.25kg          428.00

In [21]:
df.tail()

laptop_ID Company             Product            TypeName  Inches  \
907         28    Dell       Inspiron 5570            Notebook    15.6   
908       1160      HP         Spectre Pro  2 in 1 Convertible    13.3   
909         78  Lenovo  IdeaPad 320-15IKBN            Notebook    15.6   
910         23      HP              255 G6            Notebook    15.6   
911        229    Dell        Alienware 17              Gaming    17.3   

                    ScreenResolution                           Cpu   Ram  \
907                Full HD 1920x1080    Intel Core i5 8250U 1.6GHz   8GB   
908  Full HD / Touchscreen 1920x1080    Intel Core i5 6300U 2.4GHz   8GB   
909                Full HD 1920x1080    Intel Core i5 7200U 2.5GHz   8GB   
910                         1366x768  AMD E-Series E2-9000e 1.5GHz   4GB   
911      IPS Panel Full HD 1920x1080   Intel Core i7 7700HQ 2.8GHz  16GB   

                   Memory                      Gpu       OpSys  Weight  \
907             256GB SSD           AMD Radeon 530  Windows 10   2.2kg   
908             256GB SSD    Intel HD Graphics 520  Windows 10  1.48kg   
909               2TB HDD    Intel HD Graphics 620       No OS   2.2kg   
910             500GB HDD            AMD Radeon R2       No OS  1.86kg   
911  256GB SSD +  1TB HDD  Nvidia GeForce GTX 1060  Windows 10  4.42kg   

     Price_in_euros  
907          800.00  
908         1629.00  
909          519.00  
910          258.00  
911         2456.34

In [22]:
df.describe()

laptop_ID      Inches  Price_in_euros
count   912.000000  912.000000      912.000000
mean    650.312500   14.981579     1111.724090
std     382.727748    1.436719      687.959172
min       2.000000   10.100000      174.000000
25%     324.750000   14.000000      589.000000
50%     636.500000   15.600000      978.000000
75%     982.250000   15.600000     1483.942500
max    1320.000000   18.400000     6099.000000

### 2.3 Definir X e y

In [23]:
X = df.drop(['Price_in_euros'], axis=1)
y = df['Price_in_euros'].copy()
X.shape

(912, 12)

In [24]:
y.shape

(912,)

### 2.4 Dividir X_train, X_test, y_train, y_test

In [25]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.20, random_state = 42)

In [26]:
X_train

laptop_ID Company                         Product     TypeName  Inches  \
25        1118      HP                        ZBook 17  Workstation    17.3   
84         153    Dell                   Inspiron 5577       Gaming    15.6   
10         275   Apple                     MacBook Pro    Ultrabook    13.3   
342       1100      HP                   EliteBook 840     Notebook    14.0   
890        131    Dell                   Inspiron 5770     Notebook    17.3   
..         ...     ...                             ...          ...     ...   
106        578      HP  14-am079na (N3710/8GB/2TB/W10)     Notebook    14.0   
270        996  Lenovo               IdeaPad 320-15ABR     Notebook    15.6   
860        770    Dell                   Latitude 7280    Ultrabook    12.5   
435        407  Lenovo               IdeaPad 320-15IAP     Notebook    15.6   
102        418  Lenovo                  Thinkpad T470p    Ultrabook    14.0   

                       ScreenResolution                                   Cpu  \
25          IPS Panel Full HD 1920x1080           Intel Core i7 6700HQ 2.6GHz   
84                    Full HD 1920x1080           Intel Core i7 7700HQ 2.8GHz   
10   IPS Panel Retina Display 2560x1600                  Intel Core i5 2.9GHz   
342                   Full HD 1920x1080            Intel Core i5 6200U 2.3GHz   
890                   Full HD 1920x1080            Intel Core i7 8550U 1.8GHz   
..                                  ...                                   ...   
106                            1366x768  Intel Pentium Quad Core N3710 1.6GHz   
270                   Full HD 1920x1080           AMD A12-Series 9720P 3.6GHz   
860                   Full HD 1920x1080            Intel Core i7 7600U 2.8GHz   
435                            1366x768  Intel Celeron Dual Core N3350 1.1GHz   
102         IPS Panel Full HD 2560x1440           Intel Core i7 7700HQ 2.8GHz   

      Ram                Memory                      Gpu       OpSys  Weight  
25    8GB               1TB HDD       AMD FirePro W6150M   Windows 7   3.0kg  
84   16GB             512GB SSD  Nvidia GeForce GTX 1050  Windows 10  2.56kg  
10    8GB             512GB SSD  Intel Iris Graphics 550       macOS  1.37kg  
342   4GB             500GB HDD    Intel HD Graphics 520   Windows 7  1.54kg  
890  16GB  256GB SSD +  2TB HDD           AMD Radeon 530  Windows 10   2.8kg  
..    ...                   ...                      ...         ...     ...  
106   8GB               2TB HDD    Intel HD Graphics 405  Windows 10  1.94kg  
270   6GB             256GB SSD           AMD Radeon 530  Windows 10   2.2kg  
860  16GB             256GB SSD    Intel HD Graphics 620  Windows 10  1.18kg  
435   4GB               1TB HDD    Intel HD Graphics 500  Windows 10   2.2kg  
102   8GB             512GB SSD  Nvidia GeForce GT 940MX  Windows 10   1.7kg  

[729 rows x 12 columns]

In [27]:
y_train

25     2899.00
84     1249.26
10     1958.90
342    1030.99
890    1396.00
        ...   
106     389.00
270     549.00
860    1859.00
435     306.00
102    1943.00
Name: Price_in_euros, Length: 729, dtype: float64

## 3. Procesado de datos

Nuestro target es la columna `Price_in_euros`

-----------------------------------------------------------------------------------------------------------------

## 4. Modelado

### 4.1 Baseline de modelos


### 4.2 Sacar métricas, valorar los modelos

Recuerda que en la competición se va a evaluar con la métrica de ``RMSE``.

### 4.3 Optimización (up to you 🫰🏻)

-----------------------------------------------------------------

## Una vez listo el modelo, toca predecir ``test.csv``

**RECUERDA: APLICAR LAS TRANSFORMACIONES QUE HAYAS REALIZADO EN `train.csv` a `test.csv`.**


Véase:
- Estandarización/Normalización
- Eliminación de Outliers
- Eliminación de columnas
- Creación de columnas nuevas
- Gestión de valores nulos
- Y un largo etcétera de técnicas que como Data Scientist hayas considerado las mejores para tu dataset.

## 1. Carga los datos de `test.csv` para predecir.


In [28]:
X_pred = pd.read_csv("./data/test.csv")
X_pred.head()

laptop_ID Company                          Product            TypeName  \
0        209  Lenovo               Legion Y520-15IKBN              Gaming   
1       1281    Acer                   Aspire ES1-531            Notebook   
2       1168  Lenovo  V110-15ISK (i3-6006U/4GB/1TB/No            Notebook   
3       1231    Dell                    Inspiron 7579  2 in 1 Convertible   
4       1020      HP                      ProBook 640            Notebook   

   Inches                           ScreenResolution  \
0    15.6                          Full HD 1920x1080   
1    15.6                                   1366x768   
2    15.6                                   1366x768   
3    15.6  IPS Panel Full HD / Touchscreen 1920x1080   
4    14.0                          Full HD 1920x1080   

                                    Cpu   Ram     Memory  \
0           Intel Core i7 7700HQ 2.8GHz  16GB  512GB SSD   
1  Intel Celeron Dual Core N3060 1.6GHz   4GB  500GB HDD   
2            Intel Core i3 6006U 2.0GHz   4GB    1TB HDD   
3            Intel Core i5 7200U 2.5GHz   8GB  256GB SSD   
4            Intel Core i5 7200U 2.5GHz   4GB  256GB SSD   

                       Gpu       OpSys   Weight  
0  Nvidia GeForce GTX 1060       No OS    2.4kg  
1    Intel HD Graphics 400       Linux    2.4kg  
2    Intel HD Graphics 520       No OS    1.9kg  
3    Intel HD Graphics 620  Windows 10  2.191kg  
4    Intel HD Graphics 620  Windows 10   1.95kg

In [29]:
X_pred.tail()

laptop_ID  Company          Product            TypeName  Inches  \
386        820      MSI      GE72MVR 7RG              Gaming    17.3   
387        948  Toshiba  Tecra Z40-C-12X            Notebook    14.0   
388        483     Dell  Precision M5520         Workstation    15.6   
389       1017       HP      Probook 440            Notebook    14.0   
390        421     Asus     ZenBook Flip  2 in 1 Convertible    13.3   

                              ScreenResolution                          Cpu  \
386                          Full HD 1920x1080  Intel Core i7 7700HQ 2.8GHz   
387                IPS Panel Full HD 1920x1080   Intel Core i5 6200U 2.3GHz   
388                          Full HD 1920x1080  Intel Core i7 7700HQ 2.8GHz   
389                                   1366x768   Intel Core i5 7200U 2.5GHz   
390  IPS Panel Full HD / Touchscreen 1920x1080   Intel Core i5 7200U 2.5GHz   

      Ram                Memory                      Gpu       OpSys  Weight  
386  16GB  512GB SSD +  1TB HDD  Nvidia GeForce GTX 1070  Windows 10   2.9kg  
387   4GB             128GB SSD    Intel HD Graphics 520  Windows 10  1.47kg  
388   8GB             256GB SSD      Nvidia Quadro M1200  Windows 10  1.78kg  
389   4GB             500GB HDD    Intel HD Graphics 620  Windows 10  1.64kg  
390   8GB             256GB SSD    Intel HD Graphics 620  Windows 10  1.27kg

In [30]:
X_pred.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 391 entries, 0 to 390
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   laptop_ID         391 non-null    int64  
 1   Company           391 non-null    object 
 2   Product           391 non-null    object 
 3   TypeName          391 non-null    object 
 4   Inches            391 non-null    float64
 5   ScreenResolution  391 non-null    object 
 6   Cpu               391 non-null    object 
 7   Ram               391 non-null    object 
 8   Memory            391 non-null    object 
 9   Gpu               391 non-null    object 
 10  OpSys             391 non-null    object 
 11  Weight            391 non-null    object 
dtypes: float64(1), int64(1), object(10)
memory usage: 36.8+ KB


 ## 2. Replicar el procesado para ``test.csv``

In [31]:
X_pred

laptop_ID  Company                          Product            TypeName  \
0          209   Lenovo               Legion Y520-15IKBN              Gaming   
1         1281     Acer                   Aspire ES1-531            Notebook   
2         1168   Lenovo  V110-15ISK (i3-6006U/4GB/1TB/No            Notebook   
3         1231     Dell                    Inspiron 7579  2 in 1 Convertible   
4         1020       HP                      ProBook 640            Notebook   
..         ...      ...                              ...                 ...   
386        820      MSI                      GE72MVR 7RG              Gaming   
387        948  Toshiba                  Tecra Z40-C-12X            Notebook   
388        483     Dell                  Precision M5520         Workstation   
389       1017       HP                      Probook 440            Notebook   
390        421     Asus                     ZenBook Flip  2 in 1 Convertible   

     Inches                           ScreenResolution  \
0      15.6                          Full HD 1920x1080   
1      15.6                                   1366x768   
2      15.6                                   1366x768   
3      15.6  IPS Panel Full HD / Touchscreen 1920x1080   
4      14.0                          Full HD 1920x1080   
..      ...                                        ...   
386    17.3                          Full HD 1920x1080   
387    14.0                IPS Panel Full HD 1920x1080   
388    15.6                          Full HD 1920x1080   
389    14.0                                   1366x768   
390    13.3  IPS Panel Full HD / Touchscreen 1920x1080   

                                      Cpu   Ram                Memory  \
0             Intel Core i7 7700HQ 2.8GHz  16GB             512GB SSD   
1    Intel Celeron Dual Core N3060 1.6GHz   4GB             500GB HDD   
2              Intel Core i3 6006U 2.0GHz   4GB               1TB HDD   
3              Intel Core i5 7200U 2.5GHz   8GB             256GB SSD   
4              Intel Core i5 7200U 2.5GHz   4GB             256GB SSD   
..                                    ...   ...                   ...   
386           Intel Core i7 7700HQ 2.8GHz  16GB  512GB SSD +  1TB HDD   
387            Intel Core i5 6200U 2.3GHz   4GB             128GB SSD   
388           Intel Core i7 7700HQ 2.8GHz   8GB             256GB SSD   
389            Intel Core i5 7200U 2.5GHz   4GB             500GB HDD   
390            Intel Core i5 7200U 2.5GHz   8GB             256GB SSD   

                         Gpu       OpSys   Weight  
0    Nvidia GeForce GTX 1060       No OS    2.4kg  
1      Intel HD Graphics 400       Linux    2.4kg  
2      Intel HD Graphics 520       No OS    1.9kg  
3      Intel HD Graphics 620  Windows 10  2.191kg  
4      Intel HD Graphics 620  Windows 10   1.95kg  
..                       ...         ...      ...  
386  Nvidia GeForce GTX 1070  Windows 10    2.9kg  
387    Intel HD Graphics 520  Windows 10   1.47kg  
388      Nvidia Quadro M1200  Windows 10   1.78kg  
389    Intel HD Graphics 620  Windows 10   1.64kg  
390    Intel HD Graphics 620  Windows 10   1.27kg  

[391 rows x 12 columns]

In [32]:
predictions_submit = model.predict(X_pred)
predictions_submit

NameError: name 'model' is not defined

**¡OJO! ¿Por qué me da error?**

IMPORTANTE:

- SI EL ARRAY CON EL QUE HICISTEIS `.fit()` ERA DE 4 COLUMNAS, PARA `.predict()` DEBEN SER LAS MISMAS
- SI AL ARRAY CON EL QUE HICISTEIS `.fit()` LO NORMALIZASTEIS, PARA `.predict()` DEBÉIS NORMALIZARLO
- TODO IGUAL SALVO **BORRAR FILAS**, EL NÚMERO DE ROWS SE DEBE MANTENER EN ESTE SET, PUES LA PREDICCIÓN DEBE TENER **391 FILAS**, SI O SI

**Entonces, si al cargar los datos de ``train.csv`` usaste `index_col=0`, ¿tendré que hacer lo también para el `test.csv`?**

In [ ]:
# ¿Qué opináis?
# ¿Sí, no?

![wow.jpeg](attachment:wow.jpeg)

## 3. **¿Qué es lo que subirás a Kaggle?**

**Para subir a Kaggle la predicción esta tendrá que tener una forma específica.**

En este caso, la **MISMA** forma que `sample_submission.csv`.

In [33]:
sample = pd.read_csv("data/sample_submission.csv")

In [34]:
sample.head()

laptop_ID  Price_in_euros
0        209          1949.1
1       1281           805.0
2       1168          1101.0
3       1231          1293.8
4       1020          1832.6

In [35]:
sample.shape

(391, 2)

## 4. Mete tus predicciones en un dataframe llamado ``submission``.

In [36]:
#¿Cómo creamos la submission?
submission = pd.DataFrame()

In [37]:
submission.head()

Empty DataFrame
Columns: []
Index: []

In [38]:
submission.shape

(0, 0)

## 5. Pásale el CHEQUEADOR para comprobar que efectivamente está listo para subir a Kaggle.

In [39]:
def chequeador(df_to_submit):
    """
    Esta función se asegura de que tu submission tenga la forma requerida por Kaggle.

    Si es así, se guardará el dataframe en un `csv` y estará listo para subir a Kaggle.

    Si no, LEE EL MENSAJE Y HAZLE CASO.

    Si aún no:
    - apaga tu ordenador,
    - date una vuelta,
    - enciendelo otra vez,
    - abre este notebook y
    - leelo todo de nuevo.
    Todos nos merecemos una segunda oportunidad. También tú.
    """
    if df_to_submit.shape == sample.shape:
        if df_to_submit.columns.all() == sample.columns.all():
            if df_to_submit.laptop_ID.all() == sample.laptop_ID.all():
                print("You're ready to submit!")
                df_to_submit.to_csv("submission.csv", index = False) #muy importante el index = False
                urllib.request.urlretrieve("https://www.mihaileric.com/static/evaluation-meme-e0a350f278a36346e6d46b139b1d0da0-ed51e.jpg", "gfg.png")
                img = Image.open("gfg.png")
                img.show()
            else:
                print("Check the ids and try again")
        else:
            print("Check the names of the columns and try again")
    else:
        print("Check the number of rows and/or columns and try again")
        print("\nMensaje secreto del TA: No me puedo creer que después de todo este notebook hayas hecho algún cambio en las filas de `test.csv`. Lloro.")

In [40]:
chequeador(submission)

Check the number of rows and/or columns and try again

Mensaje secreto del TA: No me puedo creer que después de todo este notebook hayas hecho algún cambio en las filas de `test.csv`. Lloro.
